### DISCUSSION TITLE
Low update FPS when many notes are on screen in mania

### DISCUSSION THREAD

**Discussion by longnguyen2004** [ppy/osu#25022] [Q&A]
### Type

Performance

### Bug description

I was playing a storyboard map, which has the SV multiplier set to 0.01x to prevent turning off the storyboard. However, there are huge lag throughout the map. So I turned off the storyboard to see if it persists, which it did, so it's not due to the storyboard.

Also, this is not related to SV changes during the map, but note density in general. You can take any maps, reduce the scroll speed to minimum, and it'd exhibit the same problem (screenshot below).

### Screenshots or videos

![image](https://github.com/ppy/osu/assets/15729831/b9a60dca-2016-4845-bb4e-5d999dfc4ce8)
![image](https://github.com/ppy/osu/assets/15729831/b6b13bb5-ebf6-424c-bc40-fddb2d16a7ad)

### Version

2023.924.1

### Logs

probably not required

**Comment by peppy** [comment]
Please provide a link to the beatmap.

**Reply by longnguyen2004** [reply]
https://osu.ppy.sh/beatmapsets/1248233#mania/2594303 - loved map with 0.01x SV multiplier throughout (turn off storyboard to confirm it's not the storyboard)

But like I said, you can pick any maps, set scroll speed to 1, and the same perf degradation will occur.

**Comment by longnguyen2004** [comment]
Map: https://osu.ppy.sh/beatmapsets/533893#mania/1131026, scroll speed set to minimum to exaggerate the issue

Argon/Argon Pro
![{B60DA417-FEAD-4895-B8CB-0081EA6A2456}](https://github.com/user-attachments/assets/0c57347a-ae44-4fbf-90cf-81aa2b71ef20)

Triangles
![{5CA7B277-2AEA-443A-8BA2-09346A744443}](https://github.com/user-attachments/assets/736b7068-a763-4980-a848-3415fa685a62)

Classic
![{D25C6FD2-E20C-4407-B2A3-E086768A47CE}](https://github.com/user-attachments/assets/246bb9c2-b7f7-4553-9879-321b313b0789)

[R Skin v3.1](https://www.dropbox.com/s/mggm61kr292wq77/R%20Skin%20v3.1%20%287k%20Fade%20LN%29.osk?dl=1)
![{8826F14B-CC41-49D1-B7DB-A6AE995B6B4C}](https://github.com/user-attachments/assets/2468c348-e053-4ec7-9bce-7a98f1c19d6d)

**Reply by peppy** [reply]
Surely you don't play at this scroll speed?

**Reply by longnguyen2004** [reply]
I don't play at this scroll speed yes, it's just to exaggerate the issue. Here's another example
Map: https://osu.ppy.sh/beatmapsets/1973616#mania/4097226, scroll speed 30

Low density part (start of map), Update FPS mostly stays at the limit
![{B5576A07-7631-4C13-A31F-FF0B29DEBCA9}](https://github.com/user-attachments/assets/db02fc90-011d-4fb8-ae12-4dd7a7ef452c)

High density part (~5:20), Update FPS drops down to 180
![{08C9452F-A0F7-4B3C-9223-0878A3A1327C}](https://github.com/user-attachments/assets/acbf1d95-c742-4c24-b1a8-6cf71fdb3892)

In real gameplay (not watching a replay), these fluctuations manifest as perceivable "hitches" (dropping in smoothness) when a dense part is reached, which is highly frustrating as a player. I expect some degradation in performance when more notes are on screen, just not that much

**Reply by peppy** [reply]
Does the performance improve when you pause the game? The majority of the overhead looks to be update bound, so it's likely related to animations / movement.

**Reply by longnguyen2004** [reply]
Yes, if I pause then the performance issue goes away.

On an unrelated note, I also observed that there's a momentary stutter on the update thread *for every note hit*, which you can also see on the 2nd screenshot (very spiky frametime graph for Update), and it gets worse the longer the map goes on. Should I report it here as well or as another issue?

**Reply by peppy** [reply]
It's fine to track that here. I'll look into both.

**Reply by longnguyen2004** [reply]
Turns out it's a problem with the UR display, which recalculates the UR by [iterating through every `HitEvent`s since the start of gameplay](https://github.com/ppy/osu/blob/1e1a7293e00a5d01249c9b1d82c5a12e097bf99e/osu.Game/Rulesets/Scoring/HitEventExtensions.cs#L31), at least from what I can tell, thus the performance gets worse the more hit events there are.

While it's not mania-specific, I can make an issue if you want something to keep track of. The other issue related to note density remains however

**Reply by peppy** [reply]
In that case please make a separate issue to track, thanks!
